feat: image-builder: expandable logic of image signing (#6200)
kyma-project/test-infra: development/image-builder/README.md

* you have the `variants.yaml` file in the **same directory** as the `Dockerfile`
* your `Dockerfile` contains `ARG` directives which are named after keys in `variants.yaml`

## Image signing

> image-builder supports signing the images with pre-defined set of signing services to verify that image comes from trusted repository and has not been altered in the meantime.

Reviewer: ```suggestion
image-builder supports signing the images with a pre-defined set of signing services to verify that image comes from a trusted repository and has not been altered in the meantime.
```

> Every signing service can be enabled on repository and global levels.

Reviewer: ```suggestion
You can enable every signing service on repository and global levels.
```

> Example sign services configuration in config.yaml file:

Reviewer: ```suggestion
See the following example sign services configuration in `config.yaml` file:
```

```yaml
sign-config:
  enabled-signers:
    '*':
      - default-signify
    org/repo:
      - repo-token-notary
  signers:
    - name: default-signify
      type: notary
      config:
        endpoint: https://notary/sign
        timeout: 5m
        retry-timeout: 10s
        secret:
          path: /path/to/secret/file/signify.yaml
          type: signify
    - name: repo-token-notary
      type: notary
      config:
        endpoint: https://repo-notary/sign
        timeout: 5m
        retry-timeout: 10s
        secret:
          path: /path/to/secret/file/token
          type: token
```

> All enabled signers under `'*'` will be used globally. Additionally, if repository contains another signer configuration in the `org/repo` key, image-builder will also use this service to sign image.

Reviewer: ```suggestion
All enabled signers under `'*'` are used globally. Additionally, if a repository contains another signer configuration in the `org/repo` key, image-builder also uses this service to sign the image.
```

> When the job is running in CI (Prow), it will try to pick up the org/repo name from the default Prow variables. If binary is running outside of CI, `--repo` flag will have to be used. Otherwise, the configuration will not be used.

Reviewer: Not sure if "org/repo name" here means the same as "`org/repo` key" in the previous sentence, or whether it's something else/new.
Depending on that, I have two different suggestions:

```suggestion
If the job is running in CI (Prow), it picks up the org or repo name from the default Prow variables.
If the binary is running outside of CI, you must set the `--repo` flag. Otherwise, the configuration will not be used.
```

or

```suggestion
If the job is running in CI (Prow), it picks up the `org/repo` key from the default Prow variables.
If the binary is running outside of CI, you must set the `--repo` flag. Otherwise, the configuration will not be used.
```

Author: @NHingerl it's not a single key. It's a set of 2 env variables: `REPO_OWNER` and `REPO_NAME`. IF program works in CI, then it builds `org/repo` value from those 2 variables.

Reviewer: Then does one of my suggestions above fit, or are both not good?
Background: I want to get rid of the `/`, unless that is in the name of the key (like I assume it is in line 102).

> Currently, image-builder contains basic implementation of notary signer. If you wish to add a new signer, refer to [`sign`](./sign) package, and it's code.

Reviewer: ```suggestion
Currently, image-builder contains a basic implementation of a notary signer. If you want to add a new signer, refer to the [`sign`](./sign) package, and its code.
```

        Path to build directory context (default ".")
  -dockerfile string
        Path to Dockerfile file relative to context (default "Dockerfile")
  -env-file string
        Path to file with environment variables to be loaded in build
  -log-dir string
        Path to logs directory where GCB logs will be stored (default "/logs/artifacts")
  -name string
        Name of the image to be built
  -platform value
        Only supported with BuildKit. Platform of the image that is built
  -repo string

>         Load repository-specific configuration, eg. signing configuration.

Reviewer: ```suggestion
        Load repository-specific configuration, for example, signing configuration
```

Why did this one line get a full stop at the end, and the other lines didn't? I'm removing it for consistency.

Author: Probably because that's how it's written in the code. It's just a copy-paste from `image-builder -h`
